src/webui/torrentmetadatacache.cpp

```diff
@@ -0,0 +1,66 @@
+/*
+ * Bittorrent Client using Qt and libtorrent.
+ * Copyright (C) 2024 Thomas Piccirello <thomas@piccirello.com>
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * In addition, as a special exception, the copyright holders give permission to
+ * link this program with the OpenSSL project's "OpenSSL" library (or with
+ * modified versions of it that use the same license as the "OpenSSL" library),
+ * and distribute the linked executables. You must obey the GNU General Public
+ * License in all respects for all of the code used other than "OpenSSL".  If you
+ * modify file(s), you may extend this exception to your version of the file(s),
+ * but you are not obligated to do so. If you do not wish to do so, delete this
+ * exception statement from your version.
+ */
+
+#include "torrentmetadatacache.h"
+
+std::optional<BitTorrent::TorrentDescriptor> TorrentMetadataCache::get(const BitTorrent::InfoHash &infoHash)
+{
+    BitTorrent::TorrentDescriptor torrentDescr = m_torrentMetadata[infoHash];
+    if (isValid(torrentDescr))
+        return torrentDescr;
+    else
+        return std::nullopt;
+}
+
+bool TorrentMetadataCache::contains(const BitTorrent::InfoHash &infoHash) const
+{
+    // we don't need to check for existence in the map, the default constructed info hash won't exist in it
```

Comment:
Where in the code would the default constructed info hash be added to the map?

Reply:
> Under what circumstances would the default constructed info hash be added to the map?

Just passed in `TorrentMetadataCache::add()`.
If this is allowed by class, then you cannot assert within the class itself that it cannot be done by the caller.

Reply:
How would you prefer that I handle this? I would personally like to user an `ASSERT` given that this scenario isn't currently possible. I can also use a try/catch, however that's a bit more code (see patch below).

```patch
diff --git a/src/webui/api/torrentscontroller.cpp b/src/webui/api/torrentscontroller.cpp
index bfab97b9b..a3ff4b08c 100644
--- a/src/webui/api/torrentscontroller.cpp
+++ b/src/webui/api/torrentscontroller.cpp
@@ -50,6 +50,7 @@
 #include "base/bittorrent/torrent.h"
 #include "base/bittorrent/trackerentry.h"
 #include "base/bittorrent/trackerentrystatus.h"
+#include "base/exceptions.h"
 #include "base/interfaces/iapplication.h"
 #include "base/global.h"
 #include "base/logger.h"
@@ -1836,10 +1838,17 @@ void TorrentsController::fetchMetadataAction()
                 throw APIError(APIErrorType::BadParams, tr("Unable to download metadata for '%1'").arg(infoHash.toTorrentID().toString()));
             }
 
-            m_torrentMetadataCache.add(infoHash, sourceTorrentDescr.value());
+            try
+            {
+                m_torrentMetadataCache.add(infoHash, sourceTorrentDescr.value());
+            }
+            catch (const RuntimeError &error)
+            {
+                throw APIError(APIErrorType::BadParams, tr("Unable to fetch metadata"));
+            }
 
             setResult(serializeInfoHash(infoHash));
             setStatus(APIStatus::Async);
@@ -1881,7 +1890,14 @@ void TorrentsController::parseMetadataAction()
         if (const auto loadResult = BitTorrent::TorrentDescriptor::load(it.value()))
         {
             const BitTorrent::TorrentDescriptor &torrentDescr = loadResult.value();
-            m_torrentMetadataCache.add(torrentDescr.infoHash(), torrentDescr);
+            try
+            {
+                m_torrentMetadataCache.add(torrentDescr.infoHash(), torrentDescr);
+            }
+            catch (const RuntimeError &error)
+            {
+                throw APIError(APIErrorType::BadParams, tr("Unable to parse metadata"));
+            }
 
             const QString &fileName = it.key();
             result.insert(fileName, serializeTorrentInfo(torrentDescr));
@@ -1946,8 +1962,15 @@ void TorrentsController::onDownloadFinished(const Net::DownloadResult &result)
         {
             const BitTorrent::TorrentDescriptor &torrentDescr = loadResult.value();
             const BitTorrent::InfoHash infoHash = torrentDescr.infoHash();
-            m_torrentSourceCache.update(source, infoHash);
-            m_torrentMetadataCache.add(infoHash, torrentDescr);
+            try
+            {
+                m_torrentSourceCache.update(source, infoHash);
+                m_torrentMetadataCache.add(infoHash, torrentDescr);
+            }
+            catch (const RuntimeError &error)
+            {
+                qDebug("Unable to parse torrent from %s", qUtf8Printable(source));
+            }
         }
         else
         {
@@ -1960,15 +1983,34 @@ void TorrentsController::onDownloadFinished(const Net::DownloadResult &result)
         {
             const BitTorrent::TorrentDescriptor &torrentDescr = parseResult.value();
             const BitTorrent::InfoHash infoHash = torrentDescr.infoHash();
-            m_torrentSourceCache.update(source, infoHash);
+            try
+            {
+                m_torrentSourceCache.update(source, infoHash);
+            }
+            catch (const RuntimeError &error)
+            {
+                qDebug("Unable to fetch metadata");
+                return;
+            }
 
             if (!m_torrentMetadataCache.contains(infoHash) && !BitTorrent::Session::instance()->isKnownTorrent(infoHash))
             {
                 qDebug("Fetching metadata for %s", qUtf8Printable(infoHash.toTorrentID().toString()));
                 if (BitTorrent::Session::instance()->downloadMetadata(torrentDescr))
-                    m_torrentMetadataCache.add(infoHash, torrentDescr);
+                {
+                    try
+                    {
+                        m_torrentMetadataCache.add(infoHash, torrentDescr);
+                    }
+                    catch (const RuntimeError &error)
+                    {
+                        qDebug("Unable to fetch metadata");
+                    }
+                }
                 else [[unlikely]]
+                {
                     qDebug("Unable to fetch metadata for %s", qUtf8Printable(infoHash.toTorrentID().toString()));
+                }
             }
         }
         else {
diff --git a/src/webui/torrentmetadatacache.cpp b/src/webui/torrentmetadatacache.cpp
index 3ed78d2d4..4edb8d2b5 100644
--- a/src/webui/torrentmetadatacache.cpp
+++ b/src/webui/torrentmetadatacache.cpp
@@ -28,6 +28,8 @@
 
 #include "torrentmetadatacache.h"
 
+#include "base/exceptions.h"
+
 std::optional<BitTorrent::TorrentDescriptor> TorrentMetadataCache::get(const BitTorrent::InfoHash &infoHash)
 {
     const BitTorrent::TorrentDescriptor torrentDescr = m_torrentMetadata[infoHash];
@@ -45,6 +47,9 @@ bool TorrentMetadataCache::contains(const BitTorrent::InfoHash &infoHash) const
 
 void TorrentMetadataCache::add(const BitTorrent::InfoHash &infoHash, const BitTorrent::TorrentDescriptor &torrentDescr)
 {
+    if (infoHash == BitTorrent::InfoHash {}) [[unlikely]]
+        throw RuntimeError(tr("Cannot insert default InfoHash into cache"));
+
     m_torrentMetadata.insert(infoHash, torrentDescr);
 }
 
diff --git a/src/webui/torrentsourcecache.cpp b/src/webui/torrentsourcecache.cpp
index bb18ff826..7fea8d687 100644
--- a/src/webui/torrentsourcecache.cpp
+++ b/src/webui/torrentsourcecache.cpp
@@ -28,6 +28,8 @@
 
 #include "torrentsourcecache.h"
 
+#include "base/exceptions.h"
+
 std::optional<BitTorrent::InfoHash> TorrentSourceCache::get(const QString &source)
 {
     if (const auto iter = m_torrentSource.constFind(source); iter != m_torrentSource.constEnd())
@@ -48,6 +50,9 @@ void TorrentSourceCache::add(const QString &source)
 
 void TorrentSourceCache::update(const QString &source, const BitTorrent::InfoHash &infoHash)
 {
+    if (infoHash == BitTorrent::InfoHash {}) [[unlikely]]
+        throw RuntimeError(tr("Cannot insert default InfoHash into cache"));
+
     if (m_torrentSourcesWithoutInfoHash.contains(source))
     {
         m_torrentSource.insert(source, infoHash);
```

Reply:
> How would you prefer that I handle this? I would personally like to user an `ASSERT` given that this scenario isn't currently possible. I can also use a try/catch, however that's a bit more code (see patch below).

I still believe that all these problems only come from the fact that you are trying to put some logic in separate classes, while it would look natural in the main execution paths.

Reply:
This does not answer my questions. I'm going to end up pushing code that you don't like, we'll have a back and forth on review, and then I'll need to push different code. I'm trying to circumvent that step. Please be more pragmatic here.

Reply:
> This does not answer my questions.

What questions do you mean exactly?
If the following:

> How would you prefer that I handle this? I would personally like to user an `ASSERT` given that this scenario isn't currently possible. I can also use a try/catch, however that's a bit more code

Then I guess I answered it (maybe it was not so obvious for you to understand it):
> I still believe that all these problems only come from the fact that you are trying to put some logic in separate classes,

So I repeat.
> How would you prefer that I handle this? 

Neither one nor the other of the above suggestions. I would prefer revert these changes (related to *Cache classes).



> I'm going to end up pushing code that you don't like

There is something that I just don't like. I can say that, but I don't insist that someone necessarily redo it in my way.
But there is something that I see as clearly incorrect. I can't approve of that. And you don't have to blame me for that.

Reply:
I personally find the Cache classes easier to understand and find the logic better segmented so that it can be changed later, if needed. I prefer this approach to the prior approach of shoving all this in member variables inside TorrentsController.